### DISCUSSION TITLE
How to access a many-to-many relationship?

### DISCUSSION THREAD

**Discussion by kevinrueda** [rappasoft/laravel-livewire-tables#1386] [Q&A] [answered]
I have two models Product and Branch, both have their relationship defined with belongsToMany and I follow the Laravel conventions where the pivot table is called branch_product and it contains an is_available field. How can I use all the product fields in the datatable and add the extra is_available? Thank you. Sorry for my English

```
public function builder(): Builder
    {
        $branch = Session::get('branch')->id;
        return Product::query()->select('*')
            ->whereHas('branches', function ($query) use ($branch) {
                $query->where('branches.id', $branch);
            });
    }
```

**Comment by lrljoe** [comment, marked as answer]
Hi @kevinrueda 

Three elements here:

1. Avoid using "select" in your builder() method, as it will conflict with the internal mechanism that selects the columns for you.

Once that's resolved:

2. Keep in mind that you will need to define them as a "label" Column, e.g, this implodes the "name" of the branch, with commas:

```
Column::make('Branches')
    ->label(fn ($row) => $row->branches->pluck('name')->implode(', ')),

```


3. Using with will forcibly  load the relation for you, avoid defining a variable where you don't need to, so long as you're certain it is defined, and restrict the "with" to only return the fields you need (e.g. id, name).
```
public function builder(): Builder
{
  return Product::with(['branches:id,name' => function (Builder $query) {
      $query->where('branches.id', Session::get('branch')->id);
  }]);
}
```

**Reply by lrljoe** [reply]
No problem at all! I'd encourage joining the official Discord [https://discord.gg/jGgzP5wd65](https://discord.gg/jGgzP5wd65), as I'm fairly active on Discord, and there's plenty of experienced users who will also help out with questions.

**Reply by kevinrueda** [reply]
Hi lrljoe, the pivot table can only have one record with the same product_id and branch_id and it is already validated, I need to access a boolean field of the relationship, can I use <label> with BooleanColumn? Thank you

**Reply by lrljoe** [reply]
So do you want the BooleanColumn to be a checkbox based on whether the relationship exists, or based on a pivot value?

Either way, it will be possible to do, just might need a bit of tinkering.  Or use of a custom view, on a standard column.

The difference between a "BooleanColumn" and a "Column" is only in the view that it returns (in reality).

Feel free to drop into the Discord if that's easier for you, or share your Table Component here, and I'll do what I can to help.

**Reply by kevinrueda** [reply]
It's funny that this works perfectly for me but I have to use the boolean '**is_prepared**' field that I have in my Product model. Another thing to render the images I have to use the select method in the builder. 

```
 public function builder(): Builder
    {
        return Product::query()->select('*')
            ->whereHas('branches', function ($query) {
                $query->where('branches.id', $this->branch);
            });
    }

    public function columns(): array
    {
        return [
            Column::make("ID", "id")
                ->sortable()
                ->collapseOnMobile()
                ->deselected(),
            ImageColumn::make('Foto', 'image')
                ->location(
                    fn ($row) => asset($row->image)
                )->attributes(fn ($row) => [
                    'class' => 'img-thumbnail',
                    'alt' => $row->name,
                    'width' => '50'
                ])
                ->collapseOnMobile(),
            Column::make("Nombre", "name")
                ->sortable()
                ->searchable(),
            Column::make("Costo", "cost")
                ->sortable()
                ->searchable()
                ->collapseOnMobile(),
            Column::make("Precio", "base_price")
                ->sortable()
                ->searchable()
                ->collapseOnMobile(),
            Column::make("% IVA", "iva")
                ->sortable()
                ->searchable()
                ->collapseOnMobile(),
            BooleanColumn::make('Disponible', 'is_prepared')
                ->setCallback(function (string $value, $row) {
                    return $row->branches->where('id', $this->branch)->first()->pivot->is_available;
                }),
            Column::make('Acciones')
                ->label(fn ($row) => view(
                    'admin.tables.actions',
                    [
                        'route' => 'product',
                        'row' => $row,
                        'viewPermission' => 'view-products',
                        'editPermission' => 'edit-products',
                        'deletePermission' => 'delete-products',
                    ]
                ))
        ];
    }
```

Thanks for your time and help

**Reply by lrljoe** [reply]
For your BooleanColumn. you shouldn't need to add those restrictions.

As far as the table is concerned, you've returned:

Product with Branches (that match Session::get('branch')->id)

The Table/Component doesn't know about the other branches that exist, as they've been excluded by your earlier constraints.  So you should be able to remove that "where" statement.

If that's still working, then we can start narrowing down your builder properly, so that we end up with the field being available properly.

**Reply by kevinrueda** [reply]
When I delete the **where** it does not show me the related data, it does show the product but from another branch. 

`$row->branches>first()->pivot->is_available;`

**Reply by lrljoe** [reply]
Is branches defined as a hasMany or as a hasOne in your Model?

**Reply by kevinrueda** [reply]
```
class Product extends Model
{
    ...
    public function branches()
    {
        return $this->belongsToMany(Branch::class)->withPivot('is_available', 'have_stock', 'minimum_stock', 'current_stock');
    }
}


class Branch extends Model
{
   ...
    public function products()
    {
        return $this->belongsToMany(Product::class);
    }
}
```

**Reply by kevinrueda** [reply]
A product can belong to many branches but not more than once to one branch.

**Reply by lrljoe** [reply]
Okay, so, just so I'm clear in my head about how your relationships are formed here:

Product
hasMany Branches

But, in this case, you're restricting the Branches to only the BranchID that has been provided?

Then you want to use the PivotValue of "is_available"?

**Reply by kevinrueda** [reply]
Yes, it is restricted only by the branch ID. Product 1 can be associated with Branch 1, 2, 3 ... but it cannot be associated with a branch more than once, regardless of the other fields in the pivot table
